### ISSUE TITLE
Lock device during refresh can re-launch another

### ISSUE THREAD

**Issue by Jean-BaptisteC** [Exodus-Privacy/exodus-android-app#464]
During the first time, when the app refreshes data, if the device is locked and users unlock the device. The coroutines are re-launched.
We need to find a way to not re-launch coroutines when the device is unlocked

**Comment by Iamlooker** [comment]
Hi, if this issue is unassigned, I would like to work on this.

I looked into the code for about 30 mins and I think [this](https://github.com/Exodus-Privacy/exodus-android-app/blob/29a023ceec661fe051c2f81d655fce9ec07a2d34/app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt#L51) might be one of the root causes of the issue. But I am not sure, I think we can remove the `updateReport` from these two functions: [1](https://github.com/Exodus-Privacy/exodus-android-app/blob/29a023ceec661fe051c2f81d655fce9ec07a2d34/app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt#L203), [2](https://github.com/Exodus-Privacy/exodus-android-app/blob/29a023ceec661fe051c2f81d655fce9ec07a2d34/app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt#L211)

The mutability of the the IS_RUNNING function can be causing the issue or it could be the `LifecycleService` I can do a thorough research on it once I get a green flag 👀 

I also went through some other parts of the codebase and found a huge dependency on `MutableList` and other mutable variables (which are not using the full potential of the coroutines), as the project is leaning towards coroutines this could lead to race situations. Also the cancellation is not handled appropriately 

I think the performance can be improved significantly, but thats for another issue to handle. But just for context one thing which can be improved is, the notification shows 60/130 fetched, but none are reflected to database immediately.

**Comment by Jean-BaptisteC** [comment]
Okay, do you think `setupSwipeRefreshLayout` method is called when we unlock the device?
Don't ask to work on issue, open PR if you want :)

**Comment by Iamlooker** [comment]
Yes, I think it is been called when the fragment is resumed, but I need to test it personally

**Comment by Iamlooker** [comment]
In my recent tests I am unable to reproduce this exact issue in last few days. But I can help with improving performance in the app. I will work on this and create a PR asap

**Comment by Iamlooker** [comment]
I have requested API key via mail for testing. I have done some drastic changes and will push them once I am done testing :)

**Comment by Iamlooker** [comment]
Most changes are done, some testing is still needed

**Comment by Jean-BaptisteC** [comment]
FIY, api_key is integrated in build debug workflow

**Comment by Iamlooker** [comment]
I willtest and get be soon

**Comment by codeurimpulsif** [comment]
@Iamlooker Hey, do you still need an API key?

**Comment by Iamlooker** [comment]
> @Iamlooker Hey, do you still need an API key?

Sorry I don't think I will be able to test for few hours. I will revert back once I am back to workstation 😃

**Comment by Iamlooker** [comment]
This is the error I am getting on my debug build

```
21:30:47.054 ExodusAPIRepository      D  Attempting download of app details on com.google.android.gm.
21:30:47.287 ExodusAPIRepository      W  Failed to get app details, response code: 401. Returning emptyList.
```

I have not made any changes to ExodusAPIRepository yet

**Comment by Jean-BaptisteC** [comment]
Yes, I have same behaviour, I don't know where is the problem, maybe the API (maxrate limit?)
I suggest you to retry later

**Comment by Jean-BaptisteC** [comment]
I continue to have refresh starting when I unlock device with your changes :(

**Comment by Jean-BaptisteC** [comment]
I have find a way to reproduce:
- Start app for the first time
- Accept pop-up and notifications
- Doing nothing and wait device sleep and refresh is done
- Unlock device -> Refreshed was start for the second time
